### DISCUSSION TITLE
Upgrade request to WebSocket

### DISCUSSION THREAD

**Discussion by d-exclaimation** [hummingbird-project/hummingbird#168] [Q&A] [answered]
Great work on reaching v1.0.0. I have been interested in this project for some time now.

I wanted to ask if it is possible to programmatically upgrade a GET request to WebSocket instead of declaring a separate route for WebSocket? I am looking to do something similar to Vapor's [`req.webSocket`](https://api.vapor.codes/vapor/documentation/vapor/request/websocket(maxframesize:shouldupgrade:onupgrade:)-363b2).

I have a use case where a HTTP GET route may respond differently depending on the search parameters (query variables) and headers, one of the response may be a WebSocket switching protocol.

**Comment by adam-fowler** [comment, marked as answer]
There is no way to programmatically do it at the moment, but each WebSocket upgrade has a `shouldUpgrade` closure where you can make a decision on whether the path should upgrade to WebSockets. If you don't want to upgrade you can throw an error. You can control how that error is serialised as an HTTP reponse by conforming it to the `HBHTTPErrorResponse` protocol. If this isn't enough please add an issue to the websocket repo.

**Reply by d-exclaimation** [reply]
Thanks for the info. I'll see if I can make it work as is for now and if not, I'll open an issue.
